### PR TITLE
Add Emoji Suggestions for a More Expressive User Experience

### LINKED ISSUE #56: Emoji suggestion
### Feature description:
The user should get a suggestion of emojis after typing a special character followed by a word that matches the emoji he wants suggested.
### Example:
User types `:cat`
The suggestion bar shows 😺😸😹😻🐈...etc

Here are three screenshots as an example:
The first screenshot is taken from the Element app, it uses many keywords for each emoji.
Screenshot_20201201-224607783 (1)

The second screenshot is taken from the Discord app, it uses a single or a sequence separated by `-` as a keyword for each emoji.
Screenshot_20201201-225412880 (1)

The third screenshot is taken while using AnySoftKeyboard.
Screenshot_20201201-230918703

### PR DESCRIPTION
**Pull request by snaik20**
### Description:
- Implement `EmojiSuggestionProvider`:
  - Manages emoji suggestion tasks, including initialization and maintenance of supported emojis.
  - Generates and returns suggestions based on user input and preferences.
- Updates in `NlpManager` to Include Emoji Suggestions:
  - Adds emoji suggestions when applicable and available.
- Create Custom `EmojiSuggestionCandidate` Class:
  - Encapsulates individual emoji suggestions for seamless integration.

### Future Actions:
- Introduce Espresso tests in `Florisboard` to validate the emoji suggestion flow.

### CODE REVIEW

**Comment by patrickgold** [comment]
Quick question before I review this PR, isn't #2375 a dependency of this PR? Also what I just noticed you managed to write names into the root.txt file, the root.txt file shouldn't contain names. If you want to generate names etc. it should be named _languageCode.txt_, so for English _en.txt_, for German _de.txt_ etc and reside besides root.txt. Also the root.txt file serves as a fallback to load if other files fail.

**Comment by snaik20** [comment]
> Quick question before I review this PR, isn't #2375 a dependency of this PR? Also what I just noticed you managed to write names into the root.txt file, the root.txt file shouldn't contain names. If you want to generate names etc. it should be named _languageCode.txt_, so for English _en.txt_, for German _de.txt_ etc and reside besides root.txt. Also the root.txt file serves as a fallback to load if other files fail.

Yes both PRs should be merged for the suggestions to work. However there is no code dependency as such + that PR has a huge diff. That is why I decided to create this PR independent.

For the file generation, I used the command `./emojicon.py build en` and used the output file content into root.txt. I have not generated the names seperately. It was purely from the command. Let me know if I should have used a different command there.

**Comment by patrickgold** [comment]
Ah I see, normally the tool should have written an en.txt file, but maybe I missed that in the tool. I will review this PR then first and add root.txt manually, for the emoji suggestion file mechanism we need a different approach in the future, as each language pack provides its own localized version

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/nlp/NlpManager.kt [hunk 1]
```diff
@@ -258,16 +261,24 @@ class NlpManager(context: Context) {
         runBlocking {
             val candidates = when {
                 isSuggestionOn() -> {
-                    clipboardSuggestionProvider.suggest(
+                    emojiSuggestionProvider.suggest(
                         subtype = Subtype.DEFAULT,
```
Comment (patrickgold) [inline app/src/main/kotlin/dev/patrickgold/florisboard/ime/nlp/NlpManager.kt:265]: Must not be Subtype.DEFAULT but the currently active subtype, as emoji suggestion depends on the language of the current subtype

**Review by patrickgold — CHANGES_REQUESTED** [review]
Sorry for the late review, had a lot to do and haven't gotten to it earlier.

I have had a look at your implementation and tried testing it out but failed to successfully test your implementation. I then dived deeper into your implementation to try figure out what could go wrong and have a few points which need to be fixed.

Before that, please have a look at the https://github.com/florisboard/emojicon repo, and in there the files in `prebuilt/emoji-44.1` (freshly built using latest release CLDR 44.1 today). This directory contains all emoji language mappings the CLDR supports (159 languages + root). Files are named <language_code>.txt, which makes retrieval rather easy.

A crucial step that is missing is that these files are currently not linked into the build process, but for testing purposes I copied root.txt and en.txt into `assets/ime/media/emoji`, so this is currently a problem worth figuring out later. (You need to do this too locally)

Now what I don't understand is why no suggestions are created, I tried multiple configurations but nothing happened, also played around in your code locally but couldn't fix this, could you please have a look at this?

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/emoji/EmojiSuggestionProvider.kt [hunk 2]
```diff
@@ -0,0 +1,122 @@
+package dev.patrickgold.florisboard.ime.emoji
```
[the remark this answers is not in the document]
Comment (snaik20) [reply app/src/main/kotlin/dev/patrickgold/florisboard/ime/emoji/EmojiSuggestionProvider.kt]: Done.

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/nlp/NlpManager.kt (hunk 1 above)
Comment (snaik20) [reply app/src/main/kotlin/dev/patrickgold/florisboard/ime/nlp/NlpManager.kt]: Done.

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/emoji/EmojiSuggestionProvider.kt [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        val emojiCompatInstance = FlorisEmojiCompat.getAsFlow(emojiCompatReplaceAll).value
+        val systemFontPaint = Paint().apply { typeface = Typeface.DEFAULT }
+        val emojiPreferredSkinTone = prefs.media.emojiPreferredSkinTone.get()
+        val isEmojiSupported = { emoji: Emoji ->
+            val emojiMatch = emojiCompatInstance?.getEmojiMatch(emoji.value, emojiCompatMetadataVersion)
+            emojiMatch == EmojiCompat.EMOJI_SUPPORTED || systemFontPaint.hasGlyph(emoji.value)
+        }
+        parseRawEmojiSpecsFile(context, "ime/media/emoji/root.txt").values.flatten()
+            .filter { isEmojiSupported(it.base()) }.map { it.base(emojiPreferredSkinTone) }.toSet()
+    }
```
[the remark this answers is not in the document]
Comment (snaik20) [reply app/src/main/kotlin/dev/patrickgold/florisboard/ime/emoji/EmojiSuggestionProvider.kt]: Regarding your suggestion about the dynamic file selection based on the subtype's language, that sounds like a great approach. You mentioned a specific search order: `en-US.txt -> en.txt -> root.txt` for `en-US`

Is there a standard way to determine this order for other languages? For example, for `es_US`, would the order be `es_US.txt -> en.txt -> root.txt` or `es_US.txt -> es.txt -> en.txt -> root.txt`? Is there a module in the app that have logic implemented the logic determining search order that I could refer to?

**Comment by snaik20** [comment]
> Sorry for the late review, had a lot to do and haven't gotten to it earlier.
> 
> I have had a look at your implementation and tried testing it out but failed to successfully test your implementation. I then dived deeper into your implementation to try figure out what could go wrong and have a few points which need to be fixed.
> 
> Before that, please have a look at the https://github.com/florisboard/emojicon repo, and in there the files in `prebuilt/emoji-44.1` (freshly built using latest release CLDR 44.1 today). This directory contains all emoji language mappings the CLDR supports (159 languages + root). Files are named <language_code>.txt, which makes retrieval rather easy.
> 
> A crucial step that is missing is that these files are currently not linked into the build process, but for testing purposes I copied root.txt and en.txt into `assets/ime/media/emoji`, so this is currently a problem worth figuring out later. (You need to do this too locally)
> 
> Now what I don't understand is why no suggestions are created, I tried multiple configurations but nothing happened, also played around in your code locally but couldn't fix this, could you please have a look at this?

Thanks for your review and patience! I understand why you weren't seeing any suggestions generated.

I had updated the `root.txt` file with the emoji mappings output by emojicon tool for en.txt. But I wasn't aware of the specific search order you mentioned in your other comment (en-US.txt -> en.txt -> root.txt). This caused the issue you encountered.

Once the search order logic is implemented, this should be resolved. I have one query about the search order, I have replied on https://github.com/florisboard/florisboard/pull/2385#discussion_r1509996809 please take a look.

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/emoji/EmojiSuggestionProvider.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrickgold) [reply app/src/main/kotlin/dev/patrickgold/florisboard/ime/emoji/EmojiSuggestionProvider.kt]: You make a good point. I wouldn't really fall back to Engllish, either there are descriptions available for that language or not. However if the subtype has secondary languages defined the search order could be primary (all variants) -> secondary (all variants) -> root.txt

> Is there a module in the app that have logic implemented the logic determining search order that I could refer to?

sadly not no, the only time it is used within FlorisBoard is for the strings.xml determination however here the system does the magic without exposing the selection process nor tools

[the remark this answers is not in the document]
Comment (patrickgold) [reply app/src/main/kotlin/dev/patrickgold/florisboard/ime/emoji/EmojiSuggestionProvider.kt]: however a simple approach could be (pseudo code):

```
locale = subtype.primaryLocale
file = None
if locale.countryCode != None:
  file = tryGetEmojiFile("path/to/langCode_countryCode.txt")
if file == None:
  file = tryGetEmojiFile("path/to/langCode.txt")
if file == None:
  file = tryGetEmojiFile("path/to/root.txt")
```

The method `tryGetEmojiFile` should be cached for best results.

[the remark this answers is not in the document]
Comment (snaik20) [reply app/src/main/kotlin/dev/patrickgold/florisboard/ime/emoji/EmojiSuggestionProvider.kt]: Thanks for the review! I've added the language files and updated the logic to choose the file dynamically (PTAL).

I noticed the root.txt file seems to be missing entries for "name" and "description.", is this intentional?

File: app/src/main/assets/ime/media/emoji/en_US.txt
Comment (patrickgold) [inline app/src/main/assets/ime/media/emoji/en_US.txt:1]: these were generated in Spanish by my tool, this must be fixed by me

**Review by patrickgold — COMMENTED** [review]
Thanks a lot for the changes!

I have reviewed the code and it looks good. I then tested them out on my device and at first was super confused why there were only a few emoji's suggested, until I realized my tool generated en_US.txt in Spanish for some reason haha, that needs to be fixed.

The only thing we now need to fix is how do we get the files from the emojicon build repo into FlorisBoard? One way could be to include it as a submodule but I begin to dislike submodulles as they break easily. Another way could be to add a Python utility script in `utils/` that copies the files manually.

**Comment by polyzen** [comment]
I've been using the beta 0.4.0 releases and the stable one, now, but haven't figured out how to use emoji suggestions. I've enabled suggestions but `:cat` and the like do not provide any suggestions.

**Pull request merged**